File: src/main/java/com/windea/java/utils/YamlUtils.java
package com.windea.java.utils;

import org.yaml.snakeyaml.*;
import org.yaml.snakeyaml.constructor.Constructor;
import org.yaml.snakeyaml.representer.Representer;

import java.io.FileReader;
import java.io.FileWriter;
import java.util.*;

public class YamlUtils {
	private static LoaderOptions loaderOptions = new LoaderOptions();
	private static DumperOptions dumperOptions = new DumperOptions();

	static {
		loaderOptions.setAllowDuplicateKeys(false);
		dumperOptions.setAllowReadOnlyProperties(true);
		dumperOptions.setIndent(2);
		dumperOptions.setTimeZone(TimeZone.getTimeZone("GMT-8:00"));
		dumperOptions.setWidth(120);
		dumperOptions.setAllowUnicode(true);
	}


	private YamlUtils() {
	}


	public static void setLoaderOptions(LoaderOptions options) {
		loaderOptions = options;
	}

	public static void setDumperOptions(DumperOptions options) {
		dumperOptions = options;
	}

	public static void setTags(Map<String, String> tagMap) {
		dumperOptions.setTags(tagMap);
	}

	public static Map fromFile(String path) throws Exception {
		return fromFile(path, Map.class);
	}

	public static <T> T fromFile(String path, Class<T> type) throws Exception {
		var reader = new FileReader(path);
		T result = yaml().loadAs(reader, type);
		return result;
	}

	public static Map fromString(String string) {
		return fromString(string, Map.class);
	}

	public static <T> T fromString(String string, Class<T> type) {
		T result = yaml().loadAs(string, type);
		return result;
	}

	public static List<Object> fromFileAll(String path) throws Exception {
		var reader = new FileReader(path);
		var resultList = new ArrayList<>();
		for(var elem : yaml().loadAll(reader)) {
			resultList.add(elem);
		}
		return resultList;
	}

	public static List<Object> fromStringAll(String string) {
		var resultList = new ArrayList<>();
		for(var elem : yaml().loadAll(string)) {
			resultList.add(elem);
		}
		return resultList;
	}

	public static <T> void toFile(T data, String path) throws Exception {
		var writer = new FileWriter(path);
		yaml().dump(data, writer);
	}

	public static <T> String toString(T data) {
		return yaml().dump(data);
	}

	public static <T> void toFileAll(List<T> dataList, String path) throws Exception {
		var writer = new FileWriter(path);
		yaml().dumpAll(dataList.iterator(), writer);
	}

	public static <T> String toStringAll(List<T> dataList) {
		return yaml().dumpAll(dataList.iterator());
	}


	private static Yaml yaml() {
		var constructor = new Constructor();
		var representer = new Representer();
		return new Yaml(constructor, representer, dumperOptions, loaderOptions);
	}
}

File: src/main/kotlin/com/windea/kotlin/utils/YamlUtils.kt
package com.windea.kotlin.utils

import org.yaml.snakeyaml.*
import org.yaml.snakeyaml.constructor.Constructor
import org.yaml.snakeyaml.representer.Representer

import java.io.FileReader
import java.io.FileWriter
import java.util.*

object YamlUtils {
	private var loaderOptions = LoaderOptions()
	private var dumperOptions = DumperOptions()

	init {
		loaderOptions.isAllowDuplicateKeys = false
		dumperOptions.isAllowReadOnlyProperties = true
		dumperOptions.indent = 2
		dumperOptions.timeZone = TimeZone.getTimeZone("GMT-8:00")
		dumperOptions.width = 120
		dumperOptions.isAllowUnicode = true
	}


	fun setLoaderOptions(options: LoaderOptions) {
		loaderOptions = options
	}

	fun setDumperOptions(options: DumperOptions) {
		dumperOptions = options
	}

	fun setTags(tagMap: Map<String, String>) {
		dumperOptions.tags = tagMap
	}

	@Throws(Exception::class)
	fun fromFile(path: String): Map<*, *> {
		return fromFile(path, Map::class.java)
	}

	@Throws(Exception::class)
	fun <T> fromFile(path: String, type: Class<T>): T {
		val reader = FileReader(path)
		return yaml().loadAs(reader, type)
	}

	fun fromString(string: String): Map<*, *> {
		return fromString(string, Map::class.java)
	}

	fun <T> fromString(string: String, type: Class<T>): T {
		return yaml().loadAs(string, type)
	}

	@Throws(Exception::class)
	fun fromFileAll(path: String): List<Any> {
		val reader = FileReader(path)
		val resultList = ArrayList<Any>()
		for (elem in yaml().loadAll(reader)) {
			resultList.add(elem)
		}
		return resultList
	}

	fun fromStringAll(string: String): List<Any> {
		val resultList = ArrayList<Any>()
		for (elem in yaml().loadAll(string)) {
			resultList.add(elem)
		}
		return resultList
	}

	@Throws(Exception::class)
	fun <T> toFile(data: T, path: String) {
		val writer = FileWriter(path)
		yaml().dump(data, writer)
	}

	fun <T> toString(data: T): String {
		return yaml().dump(data)
	}

	@Throws(Exception::class)
	fun <T> toFileAll(dataList: List<T>, path: String) {
		val writer = FileWriter(path)
		yaml().dumpAll(dataList.iterator(), writer)
	}

	fun <T> toStringAll(dataList: List<T>): String {
		return yaml().dumpAll(dataList.iterator())
	}


	private fun yaml(): Yaml {
		val constructor = Constructor()
		val representer = Representer()
		return Yaml(constructor, representer, dumperOptions, loaderOptions)
	}
}
